bip-0352.mediawiki

```diff
@@ -0,0 +1,472 @@
+
+* Let ''P<sub>0</sub> = B<sub>spend</sub> + hash(outpoints_hash·a·B<sub>scan</sub> || 0)·G''
+
+Bob detects this payment by calculating ''P<sub>0</sub> = B<sub>spend</sub> + hash(outpoints_hash·b<sub>scan</sub>·A)·G'' with his online device and can spend from his cold storage signing device using ''(b<sub>spend</sub> + hash(outpoints_hash·b<sub>scan</sub>·A)) mod n'' as the private key.
+
+''' Labels '''
+
+For a single silent payment address of the form ''(B<sub>scan</sub>, B<sub>spend</sub>)'', Bob may wish to differentiate incoming payments. Naively, Bob could publish multiple silent payment addresses, but this would require him to scan for each one, which becomes prohibitively expensive. Instead, Bob can label his spend public key ''B<sub>spend</sub>'' with an integer ''m'' in the following way:
+
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + hash(b<sub>scan</sub> || m)·G'' where m is an incrementable integer starting from 1
+* Publish ''(B<sub>scan</sub>, B<sub>0</sub>)'', ''(B<sub>scan</sub>, B<sub>1</sub>)'' etc.
+
+Alice performs the tweak as before using one of the published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pairs. Bob detects the labeled payment in the following manner:
+
+* Let ''P<sub>0</sub> = B<sub>spend</sub> + hash(outpoints_hash·b<sub>scan</sub>·A || 0)·G''
+* Subtract ''P<sub>0</sub>'' from each of the transaction outputs and check if the remainder matches any of the labels (''hash(b<sub>scan</sub> || 1)·G'', hash(b<sub>scan</sub> || 2)·G'' etc.) that the wallet has previously used
+
+It is important to note that an outside observer can easily deduce that each published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pair is owned by the same entity as each published address will have ''B<sub>scan</sub>'' in common. As such, labels are not meant as a way for Bob to manage separate identities, but rather a way for Bob to determine the source of an incoming payment.
+
+''' Labels for change '''
+
+Bob can also use labels for managing his own change outputs. We reserve ''m = 0'' for this use case. This gives Bob an alternative to using BIP32 for managing change, while still allowing him to know which of his unspent outputs were change when recovering his wallet from the master key. It is important that the wallet never hands out the label with ''m = 0'' in order to ensure nobody else can create payments that are wrongly labeled as change.
+
+While the use of labels is optional, every receiving silent payments wallet should at least scan for the change label when recovering from backup in order to ensure maximum cross-compatibility.
+
+== Specification ==
+
+We use the following functions and conventions:
+
+* The constant ''n'' refers to the secp256k1 curve order, ''0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141''.
```

Comment:
_n_ is defined in BIP340 so it doesn't need to be here as well. Another option would be to put all definitions here and drop the reference to BIP340, but it is quite a lot of definitions.

Reply:
BIP340 is already referenced in "For everything not defined above, we use the notation from BIP340", so it doesn't need to be referenced again here. On the other hand, this sounds like it's saying the _curve order_ is defined in BIP340, which it's not (it's not defined anywhere, it follows from the curve parameters).

Reply:
@sipa Yes, _n_ is defined to be the curve order, however BIP340 doesn't define what the curve order itself is, and the proposed edit kind of makes it sound like it does.

Reply:
What adds to the confusion here is that $n$ is currently also used as the number of pubkeys $A_i$. And $n$ refers to the curve order only at one location in the doc, namely in the checks "if $t_k=0$ or $t_k \ge n$, fail" (twice).

I suggest getting rid of the bullet point "The constant $n$ refers to the curve order" and doing one of the following:
 - Simply remove the checks. " $t_k=0$ or $t_k \ge n$" occurs with negligible (=astronomically low) probability assuming the hash function is good, i.e., this won't occur in practice, and even if were to occur here, this won't be the end of the world if you think about it. [1] This boils down to 
 - Rephrase to "if $t_k$ is not valid tweak, i.e., if $t_k = 0$ or $t_k$ is larger or equal to the secp256k1 group order"
 - Or slightly nicer, introduce a function `valid_tweak(t)` which definition uses the words "secp256k1 group order" or simply the integer literal "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141"

(I think you also want to add an `int` call when defining $t_k$... Currently, it's a hash, so the type is a byte array, and then comparisons such as "$t_k \ge n$" don't make a lot of sense.)

[1] There tends to be a lot of bike shedding about these practically unreachable cases... Here's a hopefully convincing argument for ignoring them: All other things equal, ignoring them removes error cases and thus makes code simpler. And note that those are particularly annoying error cases: As noone knows how to reach the corresponding code paths, you can't test them, and then your line coverage suffers etc... The only inconvenient thing about my suggestion is that libsecp256k1 only provides [`secp256k1_ec_pubkey_tweak_add`](https://github.com/bitcoin-core/secp256k1/blob/d373bf6d08c82ac5496bf8103698c9f54d8d99d2/include/secp256k1.h#L763-L780), which has the checks included... :cry: (If we were to implement this function from scratch, we'd probably ignore the error case.) But that won't be an issue if we go down the proper road of having a libsecp256k1 module for silent payments.

Reply:
I prefer option 2 in that it lets the reader know this can happen but somewhat implies that it's the responsibility of the crypto library being used (not the wallet) to check that the tweak was valid.

Reply:
Updated